src/search.js

```diff
@@ -0,0 +1,125 @@
+var HMS = require('hindley-milner-search');
+
+var state = {
+  // [Node], the dom elements we will be showing/hiding when user searches
+  els: null,
+  // Node, container for results
+  container: null,
+  // Node, container for notification if there are no matches
+  containerNo: null,
+  // [Signature], the signatures as compiled by HMS
+  index: [],
+  // [Signature], most recent matches
+  matches: [],
+  // String, most recent user input
+  prevSearch: 'placeholder',
+};
+
+function map(f, x) {
+  return Array.prototype.map.call(x, f);
+}
+
+// Node -> Node
+function getGrandparent(el) {
+  return el.parentNode.parentNode;
+}
+
+// Node -> Node
+function getParent(el) {
+  return el.parentNode;
+}
+
+// Node -> Node
+function initSigResult(apiLink) {
+  var el = document.createElement('div');
+  var clone = apiLink.cloneNode(true);
+  el.appendChild(clone);
+  return el;
+}
+
+// Node -> Node -> ()
+function attachSigResult(container) {
+  return function(el) {
+    container.appendChild(el);
+  };
+}
+
+function findAndCloneSignatures(resultsContainer) {
+  var codes = document.getElementById('toc').querySelectorAll('li a code');
+  var as = map(getParent, codes);
+  var newEls = map(initSigResult, as);
+  return newEls;
+}
+
+// Node -> String
+function extractSignatureText(el) {
+  return el.innerText.trim().replace(/\s+/g, ' ');
+}
+
+// Node -> ()
+function initState(resultsContainer) {
+  state.container = document.getElementById('yes-matches');
+  state.containerNo = document.getElementById('no-matches');
+  state.els = findAndCloneSignatures(resultsContainer);
+  state.index = HMS.index(map(extractSignatureText, state.els));
+}
+
+// Node -> ()
+function clearChildren(container) {
+  while (container.childNodes.length > 0) {
+    container.removeChild(container.childNodes[0]);
+  }
+}
+function show(el) {
+  el.style.display = 'block';
+}
+
+function hide(el) {
+  el.style.display = 'none';
+}
+
+function showMatches(q, els, matches) {
+  if (matches.length === 0 && q.length > 0) {
+    show(state.containerNo);
+  }
+  else {
+    hide(state.containerNo);
+  }
```

Comment:
I don't much like the `show` and `hide` functions as they're impure. I'd prefer this mutation to appear in situ:

```javascript
state.containerNo.style.display =
  matches.length === 0 && q.length > 0 ? 'block' : 'none';
```